### DISCUSSION TITLE
application logging

### DISCUSSION THREAD

**Discussion by sattlerio** [fastapi/fastapi#7850] [Questions] [answered]
### Is your feature request related to a problem

Currently, it seems it is not possible to have an application-wide logger (handler), normally for my applications, I log into stdout and have a log format that I set on application start and it gets passed around the application.

### The solution you would like

Add in starlette an application logger, where it is possible to configure the format, level, and handlers. This could work similarly to the logger in FLASK.

### Describe alternatives you've considered

Alternative it would be nice to have in the docs a recommended way for application logging

**Comment by CorinaAio** [comment]
+1

**Comment by prostomarkeloff** [comment]
You can use the standart `logging` or `loguru` :thinking:

**Comment by sattlerio** [comment]
Hey, 
thanks for your answer. Yes indeed, but if I use the standard logging package with 2 or 3 handlers and I structure my project that the API routes are in sub-packages (like the FASTAPI) example app. Then I have to add the handlers again to the logging, which is not a problem with a logging package where I add the handlers, but the application logger from Flask and Quart is very convenient since I declare it on startup and pass it around with the app object

**Comment by sm-Fifteen** [comment, marked as answer]
[The Flask documentation on this](https://flask.palletsprojects.com/en/1.1.x/logging/) says that `flask_app.logger` is just a shortcut for `logging.getLogger('flask.app')` (or the value of `app.name` for newer versions), which the [code for this](https://github.com/pallets/flask/blob/38eb5d3b49d628785a470e2e773fc5ac82e3c8e4/src/flask/logging.py#L74) will confirm. You don't actually need to pass the logger object around, it's a singleton object, calling `logging.getLogger` whenever you need it should do.

Alternatively, you can just wrap it as a dependency:

```py
import logging

def app_logger() -> logging.Logger:
    return logging.getLogger('fastapi.sattlerio.log')

#@app.get('/')
#def root_route():
#    logger = logging.getLogger('fastapi.sattlerio.log')
#    logger.info("Hello world!")

@app.get('/')
def root_route(logger: logging.Logger = Depends(app_logger)):
    logger.info("Hello world!")
```

**Comment by sattlerio** [comment]
ok, i see. Sorry I did not check that link from the Flask docs. Since as you said it just a wrapper for logging.gelLogger it will have the same result.
I will close the ticket. Thanks for you help and sorry

**Comment by sm-Fifteen** [comment]
No worries, @sattlerio, the issues are also there to be used as questions after all. ;)

Thank you for reporting back and closing the issue.

**Comment by tiangolo** [comment]
Thanks for the help here @sm-Fifteen ! :clap: :bow: 

Thanks for reporting back and closing the issue @sattlerio  :+1:

**Comment by pcorvoh** [comment]
Hey, @sattlerio this is my program log, I don’t know if it will be helpful to you.
Minimal example:
```python
from fastapi import Request
# This is your own encapsulated logging package
from app.utils.log import Log
from config import BANNER

log = Log('request_info')

async def request_info(request: Request):
     """Get request traffic information"""
     log.info(f"{request.method} {request.url}")
     # The source of trace_id in request is middleware
     request_id = request.state.trace_id
     token = str(request.headers.get('authorization'))
     user_agent = request.headers.get('user-agent')
     body = None
     msg = "IP: %s | Request_ID: %s | User-Agent: %s | Request_Token: %s | " % (
         request.client.host, request_id, user_agent, token)
     try:
         body = await request.json()
         log.info(msg + "Request_Json: %s" % (body))
     exceptException:
         try:
             body = await request.body()
             if len(body) != 0:
                 # Have request body and record logs
                 log.info(msg + "Request_Body: %s" % (body))
             else:
                 log.info(msg)
         exceptKeyError:
             # Ignore file upload type data
             pass
-------------------------------------------------- --------------------------
# This is the middleware I use to set trace_id
class BaseMiddleware(object):
     """Middleware base class"""

     def __init__(self, app: ASGIApp) -> None:
         self.app = app

     async def __call__(
             self, scope: Scope, receive: Receive, send: Send
     ) -> None:
         if scope["type"] != "http":
             await self.app(scope, receive, send)
             return None

         request = Request(scope, receive=receive)
         request.state.trace_id = request.headers.get("x-request-id"
                                                      ) or uuid.uuid4()
         response = await self.before_request(request) or self.app
         await response(request.scope, request.receive, send)
         await self.after_request(request)

-------------------------------------------------- --------------------------
#He used it this way
from fastapi import Depends, FastAPI
app = FastAPI(
             dependencies=[Depends(request_info)],
     )
```
